packages/data/src/components/use-select/index.js

```diff
@@ -88,14 +88,24 @@ const renderQueue = createQueue();
  *
  * @return {Function}  A custom react hook.
  */
-export default function useSelect( _mapSelect, deps ) {
-	const isWithoutMapping = typeof _mapSelect !== 'function';
+export default function useSelect( mapSelect, deps ) {
+	const hasMappingFunction = 'function' === typeof mapSelect;
 
-	if ( isWithoutMapping ) {
+	// If we're recalling a store by its name or by
+	// its descriptor then we won't be caching the
+	// calls to `mapSelect` because we won't be calling it.
+	if ( ! hasMappingFunction ) {
 		deps = [];
 	}
 
-	const mapSelect = useCallback( _mapSelect, deps );
+	// Because of the "rule of hooks" we have to call `useCallback`
+	// on every invocation whether or not we have a real function
+	// for `mapSelect`. we'll create this intermediate variable to
+	// fulfill that need and then reference it with our "real"
+	// `_mapSelect` if we can.
+	const callbackMapper = useCallback( mapSelect, deps );
+	const _mapSelect = hasMappingFunction ? callbackMapper : null;
```

Comment:
To better prepare for TypeScript, the `useCallback` hook should never get `mapSelect` as a parameter if `hasMappingFunction` is `false`. Because that will not satisfy the `useCallback<T extends (...args: any[]) => any>` constraint for `T`. Better do this:
```js
const theMapSelect = hasMappingFunction ? mapSelect : noopFn;
const callbackMapper = useCallback( theMapSelect, deps );
```

Reply:
> better

my inclination was to leave it in with a type cast through `any`. as bad as it is I feel like the truly awkward part is the React hook. since you suggested it though I updated but left `_mapSelect` for the convenience of later being able to say `if ( _mapSelect ) { _mapSelect() }`